src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -176,6 +199,66 @@ public void mutuallyAuthenticate(APDU apdu) {
     respond(apdu, len, ISO7816.SW_NO_ERROR);
   }
 
+  /**
+   * Processes the LOAD_CERTS command. Copies the APDU buffer into `certs`.
+   * This function expects a DER signature and may only be called once.
+   * @param apdu the JCRE-owned APDU object.
+   */
+  public void loadCert(APDU apdu) {
+    byte[] apduBuffer = apdu.getBuffer();
+
+    final short caPubOffset = (short) ISO7816.OFFSET_CDATA;
+    final short certOffset = (short) (caPubOffset + PUBKEY_LEN);
+    final short certVerifyLen = (short)(CERTIFICATE_LEN - ECDSA_MAX_LEN);
+
+    if (idCertStatus != ID_CERTIFICATE_EMPTY) {
+      // Card cert may only be set once and never overwritten
+      ISOException.throwIt(ISO7816.SW_COMMAND_NOT_ALLOWED);
+    }
+
+    // Make sure the received certificate is appropriate length
+    if (apduBuffer[ISO7816.OFFSET_LC] != (byte) (PUBKEY_LEN + CERTIFICATE_LEN)) {
+      ISOException.throwIt(ISO7816.SW_DATA_INVALID);
+    }
+
+    JCSystem.beginTransaction();
+    // Save the certificate
+    Util.arrayCopy(apduBuffer, certOffset, idCertificate, (short) 0, CERTIFICATE_LEN);
+
+    // Lock the certificate
+    idCertStatus = ID_CERTIFICATE_LOCKED;
+    JCSystem.commitTransaction();
+  }
+  
+  /**
+   * Processes the IDENTIFY_CARD command. Returns the card public key, and a signature on the
+   * challenge salt, to prove ownership of the key.
+   * @param apdu the JCRE-owned APDU object.
+   */
+  public void identifyCard(APDU apdu) {
+    byte[] apduBuffer = apdu.getBuffer();
+
+    // Ensure the received challenge is appropriate length
+    if (apduBuffer[ISO7816.OFFSET_LC] != MessageDigest.LENGTH_SHA_256) {
+      ISOException.throwIt(ISO7816.SW_DATA_INVALID);
+    }
+
+    // Copy card ID pubKey to the response buffer
+    short responseOffset = (short) ISO7816.OFFSET_CDATA + (short) MessageDigest.LENGTH_SHA_256;
+    short off = responseOffset;
+    ECPublicKey pk = (ECPublicKey) idKeypair.getPublic();
+    short len = pk.getW(apduBuffer, off);
+    off += len;
+
+    // Sign the challenge and copy signature to response buffer
+    eccSig.init(idKeypair.getPrivate(), Signature.MODE_SIGN);
+    len = eccSig.signPreComputedHash(apduBuffer, (short) ISO7816.OFFSET_CDATA, MessageDigest.LENGTH_SHA_256, apduBuffer, off);
```

Comment:
Lol wait... so you're parameterized your signature to take SHA256 hashes and you still have to pass the size of the message here?

Note that I've run into issues in the past where it was hashing the message before signing it. I don't think you will encounter that here given the name of the function, but just be aware

Reply:
Yeah, so much fun :)

Good to know